app/lib/search/index_simple.dart

```diff
@@ -168,6 +161,80 @@ class SimplePackageIndex implements PackageIndex {
       value: (String url) => _documents[url].popularity * 100,
     );
   }
+
+  Score _searchText(String text, String packagePrefix) {
+    Score textScore;
+    if (text != null && text.isNotEmpty) {
+      textScore = new Score()
+        ..addValues(_nameIndex.search(text), 0.82)
+        ..addValues(_descrIndex.search(text), 0.12)
+        ..addValues(_readmeIndex.search(text), 0.06);
+      textScore.removeLowScores(0.05);
+    } else if (packagePrefix != null) {
+      textScore = new Score()..addValues(_nameIndex.search(packagePrefix), 0.8);
+    }
+    return textScore;
+  }
+
+  List<PackageScore> _flattenFromValues(Map<String, double> values) {
+    final List<PackageScore> list = values.keys
+        .map((url) => new PackageScore(
+              url: url,
+              package: _documents[url].package,
+              score: values[url],
+            ))
+        .toList();
+    list.sort((a, b) {
+      final int scoreCompare = -a.score.compareTo(b.score);
+      if (scoreCompare != 0) return scoreCompare;
+      // if two packages got the same score, order by last updated
+      return _compareUpdated(_documents[a.url], _documents[b.url]);
+    });
+    return list;
+  }
+
+  List<PackageScore> _flattenFromUrls(
+      Set<String> urls, int compare(PackageDocument a, PackageDocument b)) {
+    final List<PackageScore> list = urls
+        .map((url) =>
+            new PackageScore(url: url, package: _documents[url].package))
+        .toList();
+    list.sort((a, b) => compare(_documents[a.url], _documents[b.url]));
+    return list;
+  }
+
+  int _compareUpdated(PackageDocument a, PackageDocument b) {
+    if (a.updated == null) return -1;
+    if (b.updated == null) return 1;
+    return -a.updated.compareTo(b.updated);
+  }
+}
+
+class Score {
+  final Map<String, double> values = <String, double>{};
+
+  Iterable<String> getKeys() => values.keys;
+
+  void addValues(Map<String, double> newValues, double weight) {
+    if (newValues == null) return;
+    newValues.forEach((String key, double score) {
+      if (score != null) {
+        final double prev = values[key] ?? 0.0;
+        values[key] = prev + score * weight;
+      }
+    });
+  }
+
+  void removeWhere(bool keyCondition(String key)) {
+    final Set<String> keysToRemove = values.keys.where(keyCondition).toSet();
+    keysToRemove.forEach(values.remove);
+  }
+
+  void removeLowScores(double fraction) {
+    final double maxValue = values.values.fold(0.0, max);
```

Comment:
I think this is a big weird: Basically if I search for "foobar" and none of the packages are relevant, though our character n-grams returns some low-score results, we find the max here, and someone passes like `fraction = 0.9` (meaning should be in top 10%) then we get results, even though none are relevant.

I think a better way of doing this is to always ensure that if a package is perfect match, then it's score needs to be `1.0` then you can just say

```dart
removeWhere((key) => values[key] < 0.9);
```

Then it's even possible to have optimizations which pushes this filter down to the things calling `addValues` so you can avoid work if you know already it can't pass the filter.

Reply:
This may happen automatically, because we weight longer N-grams and prefixes more, and if there is a 6+ character match, we will most likely remove the low-quality matches anyway.
I'll experiment with this, and do a follow-up in a subsequent PR.

Reply:
Maybe slightly said in another way: If a search query has **only results with poor scores**, we should **display none of them** - not even the highest scored one.

Reply:
Yeah, that makes sense. I've added a low-score filter, but it is very conservative now. I would still experiment with it more, because it is not always the package name that people are searching for, and the rules around it may become complex.